pennylane/ops/op_math/prod.py

```diff
@@ -247,30 +242,41 @@ def decomposition(self):
     def matrix(self, wire_order=None):
         """Representation of the operator as a matrix in the computational basis."""
 
-        def mats_gen():
-            for ops in self.overlapping_ops:
-                if len(ops) == 1:
-                    yield (
-                        qml.matrix(ops[0])
-                        if isinstance(ops[0], qml.Hamiltonian)
-                        else ops[0].matrix()
-                    )
+        mats = []
+        for ops in self.overlapping_ops:
+            if len(ops) == 1:
+                batched = ops[0].batch_size is not None
+                mat = qml.matrix(ops[0]) if isinstance(ops[0], qml.Hamiltonian) else ops[0].matrix()
+                if self.batch_size and not batched:
+                    mats.append(qml.math.stack([mat for _ in range(self.batch_size)]))
```

Comment:
Could we get away with memorizing the information `batched` and using it later on, instead of copying the matrix? :thinking:

Reply:
And another thought: Do we need to `stack` here, if we anyways manually iterate over the stacked axis later on (in the `reduce(math.kron...` line)?

Reply:
Nice catch!! Created a list containing bools that specify if the matrix is batched or not, and use it later to re-use the matrix.

Reply:
Nice! :100: